File: services/langchain/handler.py
"""Lambda Handler - LangChain Service

AWS Lambda Container Image用のハンドラー。
Lambda Function URL または API Gateway から呼び出される。
AgentCore Runtimeへのプロキシ機能も提供。
"""

import base64
import json
import logging
import os
from typing import Any

import boto3

from agent import ChatRequest, chat, health, info

# ロギング設定
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# AgentCore client
bedrock_agentcore_client = None


def get_agentcore_client():
    """AgentCore クライアントを取得（遅延初期化）"""
    global bedrock_agentcore_client
    if bedrock_agentcore_client is None:
        # Lambda環境ではAWS_REGIONが自動設定、BEDROCK_REGIONも使用可能
        region = os.getenv("BEDROCK_REGION") or os.getenv("AWS_REGION", "us-east-1")
        bedrock_agentcore_client = boto3.client(
            "bedrock-agentcore",
            region_name=region,
        )
    return bedrock_agentcore_client


def invoke_agentcore(
    runtime_arn: str,
    instruction: str,
    endpoint_id: str | None = None,
) -> dict[str, Any]:
    """AgentCore Runtime を呼び出す"""
    import time

    start_time = time.time()

    client = get_agentcore_client()

    # ペイロード作成 - CLIと同じ形式
    payload = {"input": {"prompt": instruction}}
    payload_json = json.dumps(payload)
    # SDK は bytes を期待する場合がある
    payload_bytes = payload_json.encode("utf-8")

    logger.info(f"Invoking AgentCore: {runtime_arn}")
    logger.info(f"Payload (raw): {payload_json}")

    try:
        # SDK呼び出し - bytes形式で渡す
        response = client.invoke_agent_runtime(
            agentRuntimeArn=runtime_arn,
            payload=payload_bytes,
        )

        logger.info(f"Response keys: {list(response.keys())}")

        # SDK レスポンスからデータを読み取り
        # SDK は 'response' キーを使用（CLI は outfile に body を書き込む）
        body = response.get("response") or response.get("body")
        if body is None:
            # response全体からメタデータを除いて探す
            for key in ["output", "result", "data"]:
                if key in response:
                    body = response[key]
                    break

        if body is None:
            logger.error(f"Full response: {response}")
            raise ValueError(f"No body in AgentCore response. Keys: {list(response.keys())}")

        # StreamingBody の場合は read() でバイトを取得
        if hasattr(body, "read"):
            body_bytes = body.read()
            output_data = json.loads(body_bytes.decode("utf-8"))
        elif isinstance(body, bytes):
            output_data = json.loads(body.decode("utf-8"))
        elif isinstance(body, str):
            output_data = json.loads(body)
        elif isinstance(body, dict):
            output_data = body
        else:
            output_data = {"raw": str(body)}

        latency_ms = int((time.time() - start_time) * 1000)

        # メタデータに latency を追加
        if "metadata" not in output_data:
            output_data["metadata"] = {}
        output_data["metadata"]["latency_ms"] = latency_ms
        output_data["metadata"]["runtime_arn"] = runtime_arn

        logger.info(f"AgentCore response received in {latency_ms}ms")

        return output_data

    except Exception as e:
        logger.error(f"AgentCore SDK call failed: {type(e).__name__}: {e}")
        raise


def lambda_handler(event: dict[str, Any], context: Any) -> dict[str, Any]:
    """Lambda Handler

    Lambda Function URL からのリクエストを処理。

    エンドポイント:
    - POST /api/v1/chat: チャット実行
    - POST /api/v1/chat/tools: ツール付きチャット
    - GET /api/v1/health: ヘルスチェック
    - GET /api/v1/info: サービス情報
    """
    try:
        # リクエスト情報を抽出
        http_method = event.get("requestContext", {}).get("http", {}).get("method", "GET")
        path = event.get("rawPath", "/")

        logger.info(f"Request: {http_method} {path}")

        # CORS headers
        headers = {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
            "Access-Control-Allow-Headers": "Content-Type, Authorization",
        }

        # OPTIONS (CORS preflight)
        if http_method == "OPTIONS":
            return {
                "statusCode": 200,
                "headers": headers,
                "body": "",
            }

        # ルーティング
        if path == "/api/v1/health" and http_method == "GET":
            response = health()
            return {
                "statusCode": 200,
                "headers": headers,
                "body": response.model_dump_json(),
            }

        elif path == "/api/v1/info" and http_method == "GET":
            response = info()
            return {
                "statusCode": 200,
                "headers": headers,
                "body": response.model_dump_json(),
            }

        elif path == "/api/v1/chat" and http_method == "POST":
            body = json.loads(event.get("body", "{}"))

            # AgentCoreへのプロキシ
            target_service = body.get("target_service")
            if target_service == "agentcore":
                runtime_arn = body.get(
                    "agentcore_runtime_arn",
                    "arn:aws:bedrock-agentcore:us-east-1:226484346947:runtime/agentcore_strands_dev-sSCXyh2bVa",
                )
                endpoint_id = body.get("agentcore_endpoint_id", "agentcore_strands_dev_endpoint")

                try:
                    agentcore_response = invoke_agentcore(
                        runtime_arn=runtime_arn,
                        instruction=body.get("instruction", ""),
                        endpoint_id=endpoint_id,
                    )

                    # AgentCoreレスポンスをフォーマット
                    output = agentcore_response.get("output", {})
                    message = output.get("message", {})
                    content_list = message.get("content", [])
                    content_text = ""
                    for item in content_list:
                        if isinstance(item, dict) and "text" in item:
                            content_text += item["text"]

                    formatted_response = {
                        "response_id": agentcore_response.get("metadata", {}).get("request_id", "agentcore-response"),
                        "content": content_text,
                        "tool_calls": output.get("tool_calls"),
                        "latency_ms": agentcore_response.get("metadata", {}).get("latency_ms", 0),
                        "metadata": {
                            "service": "agentcore",
                            "framework": "strands-agents",
                            "model_id": agentcore_response.get("metadata", {}).get("model_id", "unknown"),
                            "region": os.getenv("AWS_REGION", "us-east-1"),
                        },
                    }

                    return {
                        "statusCode": 200,
                        "headers": headers,
                        "body": json.dumps(formatted_response),
                    }
                except Exception as e:
                    logger.exception(f"AgentCore invoke error: {e}")
                    return {
                        "statusCode": 500,
                        "headers": headers,
                        "body": json.dumps({"error": "AgentCore Error", "details": str(e)}),
                    }

            # 通常のLangChain処理
            request = ChatRequest(
                instruction=body.get("instruction", ""),
                session_id=body.get("session_id"),
                use_tools=body.get("use_tools", False),
            )
            response = chat(request)
            return {
                "statusCode": 200,
                "headers": headers,
                "body": response.model_dump_json(),
            }

        elif path == "/api/v1/chat/tools" and http_method == "POST":
            body = json.loads(event.get("body", "{}"))
            request = ChatRequest(
                instruction=body.get("instruction", ""),
                session_id=body.get("session_id"),
                use_tools=True,  # ツール使用を強制
            )
            response = chat(request)
            return {
                "statusCode": 200,
                "headers": headers,
                "body": response.model_dump_json(),
            }

        # ルート
        elif path == "/" and http_method == "GET":
            return {
                "statusCode": 200,
                "headers": headers,
                "body": json.dumps(
                    {
                        "service": "LangChain Service",
                        "version": "1.0.0",
                        "endpoints": [
                            "/api/v1/chat",
                            "/api/v1/chat/tools",
                            "/api/v1/health",
                            "/api/v1/info",
                        ],
                    }
                ),
            }

        # 404 Not Found
        else:
            return {
                "statusCode": 404,
                "headers": headers,
                "body": json.dumps({"error": "Not Found", "path": path}),
            }

    except json.JSONDecodeError as e:
        logger.error(f"JSON decode error: {e}")
        return {
            "statusCode": 400,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": "Invalid JSON", "details": str(e)}),
        }
    except Exception as e:
        logger.exception(f"Handler error: {e}")
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": "Internal Server Error", "details": str(e)}),
        }


# ローカルテスト用
if __name__ == "__main__":
    # テストイベント
    test_event = {
        "requestContext": {"http": {"method": "POST"}},
        "rawPath": "/api/v1/chat",
        "body": json.dumps({"instruction": "東京の天気を教えて", "use_tools": True}),
    }

    result = lambda_handler(test_event, None)
    print(json.dumps(json.loads(result["body"]), indent=2, ensure_ascii=False))

File: frontend/amplify/functions/service-api/handler.ts
import type { Handler } from 'aws-lambda';
import {
  BedrockRuntimeClient,
  ConverseCommand,
} from '@aws-sdk/client-bedrock-runtime';

const client = new BedrockRuntimeClient({ region: 'us-east-1' });

interface ServiceExecuteRequest {
  instruction: string;
  agent_type: 'strands' | 'langchain';
  tools?: Record<string, unknown>[];
  use_memory?: boolean;
  session_id?: string;
}

interface ServiceExecuteResponse {
  response_id: string;
  content: string;
  tool_calls?: Record<string, unknown>[];
  latency_ms: number;
  metadata?: Record<string, unknown>;
  framework_features?: string[];
}

interface LambdaEvent {
  httpMethod?: string;
  requestContext?: {
    http?: {
      method: string;
      path: string;
    };
  };
  rawPath?: string;
  path?: string;
  body?: string;
  isBase64Encoded?: boolean;
}

interface LambdaResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

/**
 * Generate ULID-like ID
 */
function generateId(): string {
  const timestamp = Date.now().toString(36);
  const random = Math.random().toString(36).substring(2, 12);
  return `${timestamp}${random}`.toUpperCase();
}

/**
 * Call Bedrock Converse API
 */
async function callBedrock(prompt: string): Promise<string> {
  const modelId = process.env.BEDROCK_MODEL_ID || 'anthropic.claude-3-5-sonnet-20241022-v2:0';
  
  try {
    const command = new ConverseCommand({
      modelId,
      messages: [
        {
          role: 'user',
          content: [{ text: prompt }],
        },
      ],
      inferenceConfig: {
        maxTokens: 1024,
        temperature: 0.7,
      },
    });

    const response = await client.send(command);
    
    const content = response.output?.message?.content?.[0];
    if (content && 'text' in content && content.text) {
      return content.text;
    }
    
    return 'No response generated.';
  } catch (error) {
    console.error('Bedrock API error:', error);
    return `Error calling Bedrock: ${error instanceof Error ? error.message : 'Unknown error'}`;
  }
}

/**
 * Get framework-specific features
 */
function getFrameworkFeatures(agentType: 'strands' | 'langchain'): string[] {
  if (agentType === 'strands') {
    return [
      'bedrock_native_integration',
      'agentcore_memory_api',
      'prompt_caching',
      'tool_caching',
      'automatic_tool_loop',
      'guardrails',
    ];
  }
  return [
    'langgraph_state_management',
    'checkpointing',
    'tool_node_automation',
    'multi_provider_support',
    'time_travel_debugging',
  ];
}

/**
 * Main handler
 */
export const handler: Handler<LambdaEvent, LambdaResponse> = async (event) => {
  // CORS headers
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type, Authorization',
  };

  // Get HTTP method and path (works with both API Gateway and Function URL)
  const method = event.requestContext?.http?.method || event.httpMethod || 'GET';
  const path = event.requestContext?.http?.path || event.rawPath || event.path || '/';

  console.log(`Request: ${method} ${path}`);

  // Handle OPTIONS (CORS preflight)
  if (method === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  try {
    // Parse body
    let body: ServiceExecuteRequest | undefined;
    if (event.body) {
      const rawBody = event.isBase64Encoded 
        ? Buffer.from(event.body, 'base64').toString('utf-8')
        : event.body;
      body = JSON.parse(rawBody);
    }

    // Health check
    if (path.endsWith('/health') || path === '/') {
      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({ status: 'ok', version: '1.0.0', path }),
      };
    }

    // Service info
    if (path.includes('/info') && method === 'GET') {
      const serviceName = path.split('/services/')[1]?.split('/')[0] || 'runtime';
      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({
          service: serviceName,
          description: `${serviceName} service for AgentCore vs LangChain comparison`,
          supported_agents: ['strands', 'langchain'],
          strands_features: getFrameworkFeatures('strands'),
          langchain_features: getFrameworkFeatures('langchain'),
          comparison_available: true,
        }),
      };
    }

    // Service execute
    if (path.includes('/execute') && method === 'POST' && body) {
      const startTime = Date.now();

      // Simulate agent-specific behavior
      let systemPrompt = '';
      if (body.agent_type === 'strands') {
        systemPrompt = 'You are an AI assistant powered by AWS Bedrock AgentCore (Strands Agents). ';
      } else {
        systemPrompt = 'You are an AI assistant powered by LangChain/LangGraph. ';
      }

      const fullPrompt = `${systemPrompt}\n\nUser: ${body.instruction}`;
      const content = await callBedrock(fullPrompt);
      
      const latencyMs = Date.now() - startTime;

      const response: ServiceExecuteResponse = {
        response_id: generateId(),
        content,
        latency_ms: latencyMs,
        metadata: {
          service: path.split('/services/')[1]?.split('/')[0] || 'runtime',
          agent_type: body.agent_type,
          model_id: process.env.BEDROCK_MODEL_ID,
        },
        framework_features: getFrameworkFeatures(body.agent_type),
      };

      return {
        statusCode: 200,
        headers,
        body: JSON.stringify(response),
      };
    }

    // Service compare (run both)
    if (path.includes('/compare') && method === 'POST' && body) {
      // Run Strands
      const strandsStart = Date.now();
      const strandsPrompt = `You are an AI assistant powered by AWS Bedrock AgentCore (Strands Agents).\n\nUser: ${body.instruction}`;
      const strandsContent = await callBedrock(strandsPrompt);
      const strandsLatency = Date.now() - strandsStart;

      // Run LangChain
      const langchainStart = Date.now();
      const langchainPrompt = `You are an AI assistant powered by LangChain/LangGraph.\n\nUser: ${body.instruction}`;
      const langchainContent = await callBedrock(langchainPrompt);
      const langchainLatency = Date.now() - langchainStart;

      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({
          strands_result: {
            response_id: generateId(),
            content: strandsContent,
            latency_ms: strandsLatency,
            framework_features: getFrameworkFeatures('strands'),
          },
          langchain_result: {
            response_id: generateId(),
            content: langchainContent,
            latency_ms: langchainLatency,
            framework_features: getFrameworkFeatures('langchain'),
          },
          comparison: {
            latency_diff_ms: strandsLatency - langchainLatency,
            faster_framework: strandsLatency < langchainLatency ? 'strands' : 'langchain',
          },
        }),
      };
    }

    // Not found
    return {
      statusCode: 404,
      headers,
      body: JSON.stringify({ error: 'Not found', path, method }),
    };
  } catch (error) {
    console.error('Error:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        error: error instanceof Error ? error.message : 'Unknown error',
      }),
    };
  }
};
